test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java

```diff
@@ -158,5 +162,69 @@ public void testResolveReleasedVersionsForUnstableBranch() {
         assertEquals(Arrays.asList(TestUnstableBranch.V_5_3_2, TestUnstableBranch.V_5_4_0), unreleased);
     }
 
-    // TODO add a test that compares gradle and VersionUtils.java in a followup
+    /**
+     * Tests that {@link Version#minimumCompatibilityVersion()} and {@link VersionUtils#allReleasedVersions()}
+     * agree with the list of wire and index compatible versions we build in gradle.
+     */
+    public void testGradleVersionsMatchVerionUtils() {
+        // First check the index compatible versions
+        VersionsFromProperty indexCompatible = new VersionsFromProperty("tests.gradle_index_compat_versions");
+
+        List<Version> released = VersionUtils.allReleasedVersions().stream()
+                /* We skip alphas, betas, and the like in gradle because they don't have
+                 * backwards compatibility guarantees even though they are technically
+                 * released. */
+                .filter(Version::isRelease)
+                .collect(toList());
+        List<String> releasedIndexCompatible = released.stream()
+                .map(Object::toString)
+                .collect(toList());
+        assertEquals(releasedIndexCompatible, indexCompatible.released);
+
+        List<String> unreleasedIndexCompatible = VersionUtils.allUnreleasedVersions().stream()
+                .map(Object::toString)
+                .collect(toList());
+        assertEquals(unreleasedIndexCompatible, indexCompatible.unreleased);
+
+        // Now the wire compatible versions
+        VersionsFromProperty wireCompatible = new VersionsFromProperty("tests.gradle_wire_compat_versions");
+
+        Version minimumCompatibleVersion = Version.CURRENT.minimumCompatibilityVersion();
+        List<String> releasedWireCompatible = released.stream()
+                .filter(v -> v.onOrAfter(minimumCompatibleVersion))
+                .map(Object::toString)
+                .collect(toList());
+        assertEquals(releasedWireCompatible, wireCompatible.released);
+
+        List<String> unreleasedWireCompatible = VersionUtils.allUnreleasedVersions().stream()
+                .filter(v -> v.onOrAfter(minimumCompatibleVersion))
+                .map(Object::toString)
+                .collect(toList());
+        assertEquals(unreleasedWireCompatible, wireCompatible.unreleased);
+    }
+
+    /**
+     * Read a versions system property as set by gradle into a tuple of {@code (releasedVersion, unreleasedVersion)}.
+     */
+    private class VersionsFromProperty {
+        private final List<String> released = new ArrayList<>();
+        private final List<String> unreleased = new ArrayList<>();
+
+        public VersionsFromProperty(String property) {
+            String versions = System.getProperty(property);
+            assertNotNull("Couldn't find [" + property + "]. Gradle should set these before running the tests.", versions);
+            assertThat(versions, startsWith("["));
```

Comment:
Why pass them in as an array and have to do extra parsing here, instead of a comma separated list?